network/test/epochtransition_test.go

```diff
@@ -31,138 +34,258 @@ import (
+type testNodeList []testNode
+
+func (t testNodeList) ids() flow.IdentityList {
+	ids := make(flow.IdentityList, len(t))
+	for i, node := range t {
+		ids[i] = node.id
+	}
+	return ids
+}
+
+func (t testNodeList) engines() []*MeshEngine {
+	engs := make([]*MeshEngine, len(t))
+	for i, node := range t {
+		engs[i] = node.engine
+	}
+	return engs
+}
+
+func (t testNodeList) idRefreshers() []*p2p.NodeIDRefresher {
+	idRefreshers := make([]*p2p.NodeIDRefresher, len(t))
+	for i, node := range t {
+		idRefreshers[i] = node.idRefresher
+	}
+	return idRefreshers
 }
 
 func TestEpochTransitionTestSuite(t *testing.T) {
 	suite.Run(t, new(MutableIdentityTableSuite))
 }
 
 func (suite *MutableIdentityTableSuite) SetupTest() {
+	suite.testNodes = nil
 	rand.Seed(time.Now().UnixNano())
 	nodeCount := 10
-	suite.logger = zerolog.New(os.Stderr).Level(zerolog.ErrorLevel)
+	suite.logger = zerolog.New(os.Stderr).Level(zerolog.DebugLevel)
 	log.SetAllLoggers(log.LevelError)
 
-	// create ids
-	ids, mws := GenerateIDsAndMiddlewares(suite.T(), nodeCount, !DryRun, suite.logger)
-	suite.ids = ids
-	suite.mws = mws
+	suite.setupStateMock()
+	suite.addNode(nodeCount)
+
+	// simulate a start of an epoch by signaling a change in the identity table
+	suite.signalIdentityChanged()
+
+	// wait for two lip2p heatbeats for the nodes to discover each other and form the mesh
+	time.Sleep(2 * time.Second)
+}
 
-	// setup state related mocks
+// TearDownTest closes all the networks within a specified timeout
+func (suite *MutableIdentityTableSuite) TearDownTest() {
+	nets := make([]*p2p.Network, 0, len(suite.testNodes)+len(suite.removedTestNodes))
+	for _, n := range suite.testNodes {
+		nets = append(nets, n.net)
+	}
+	for _, n := range suite.removedTestNodes {
+		nets = append(nets, n.net)
+	}
+	stopNetworks(suite.T(), nets, 3*time.Second)
+}
+
+// setupStateMock setup state related mocks (all networks share the same state mock)
+func (suite *MutableIdentityTableSuite) setupStateMock() {
 	final := unittest.BlockHeaderFixture()
 	suite.state = new(mockprotocol.State)
 	suite.snapshot = new(mockprotocol.Snapshot)
 	suite.snapshot.On("Head").Return(&final, nil)
 	suite.snapshot.On("Phase").Return(flow.EpochPhaseCommitted, nil)
+	// return all the current list of ids for the state.Final.Identities call made by the network
 	suite.snapshot.On("Identities", mock.Anything).Return(
-		func(flow.IdentityFilter) flow.IdentityList { return suite.ids },
+		func(flow.IdentityFilter) flow.IdentityList {
+			return suite.testNodes.ids()
+		},
 		func(flow.IdentityFilter) error { return nil })
 	suite.state.On("Final").Return(suite.snapshot, nil)
+}
 
-	// all nodes use the same state mock
-	states := make([]*mockprotocol.State, nodeCount)
-	for i := 0; i < nodeCount; i++ {
-		states[i] = suite.state
+// addNode creates count many new nodes and appends them to the suite state variables
+func (suite *MutableIdentityTableSuite) addNode(count int) {
```

Comment:
```suggestion
func (suite *MutableIdentityTableSuite) addNodes(count int) {
```
if we're adding multiple nodes